lib/debuglet.js

```diff
@@ -434,3 +441,53 @@ Debuglet.mapSubtract = function mapSubtract(A, B) {
   return removed;
 };
 
+/**
+ * Formats the message base with placeholders `$0`, `$1`, etc
+ * by substituting the provided expressions. If more expressions
+ * are given than placeholders extra expressions are dropped.
+ */
+Debuglet.format = function(base, exprs) {
```

Comment:
I think this code would be *much* simpler if you wrote it as a simple nested loop scanning the input from left to right – no need to tokenize.